src/main/kotlin/com/terraformation/backend/accelerator/db/ReportStore.kt

```diff
@@ -349,6 +350,38 @@ class ReportStore(
     }
   }
 
+  /**
+   * Fetch upcoming reports and notify that they are due. Also marks the report notification time.
+   */
+  fun notifyUpcomingReports(): Int {
+    requirePermissions { notifyUpcomingReports() }
+
+    val today = LocalDate.ofInstant(clock.instant(), ZoneId.systemDefault())
+
+    return with(REPORTS) {
+      dslContext.transactionResult { _ ->
+        // Lock these rows to prevent multiple notifications being sent per report
+        val reportIds =
+            dslContext
+                .select(ID)
+                .from(this)
+                .where(STATUS_ID.eq(ReportStatus.NotSubmitted))
+                .and(UPCOMING_NOTIFICATION_SENT_TIME.isNull)
+                .and(END_DATE.lessOrEqual(today.plusDays(15)))
+                .forUpdate()
```

Comment:
Use `.forUpdate().skipLocked()` so that if this runs on multiple server hosts at once, the first one to run the query will grab the lock and the others won't block waiting for it.